### DISCUSSION TITLE
How do I use multiple variables on columns with high cardinality?

### DISCUSSION THREAD

**Discussion by mitulk** [grafana/grafana#44615] [Dashboards, Panels & Visualizations]
My table has columns with high cardinality, e.g. pageViews table in Azure Monitor Application insights (of course, this applies to other data sources as well).

For example, consider columns C1 and C2. I have variables v1 and v2 setup against these columns, and a panel that uses these variables to filter the data in that panel. 

Once I pick a value for v1, I would like v2 to only show those values that make sense for v1. This is doable by adding a v1 based filter to the query for v2. This is helpful, as there is a high number of possible values for v2 otherwise. Selecting irrelevant values for v2 could result in an empty panel.

However, this dashboard now has a unidirectional dependency - I have to pick v1 first and only then can I choose a relevant value for v2. There isn't a way to do the reverse on this dashboard - choose a value for v2 and then choose a relevant value for v1. 

A quick search of the Grafana issues/ source code shows that circular dependency is not supported on variables:
- https://github.com/grafana/grafana/issues/8583, https://github.com/grafana/grafana/issues/12743
- [public/app/core/utils/dag.ts](https://github.com/grafana/grafana/blob/main/public/app/core/utils/dag.ts)

It makes sense that circular dependencies are not supported. One could argue that bidirectional dependency != circular dependency. 

The code currently doesn't differentiate between the two, nor does it prevent it. For example, I can craft bidirectional variable queries against Azure Data Explorer; causing an infinite loop of continuous queries.

Here's a sample dashboard json:
```
{
  "annotations": {
    "list": []
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 1,
  "iteration": 1643406930029,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 24,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single",
          "sort": "none"
        }
      },
      "pluginVersion": "7.5.0",
      "targets": [
        {
          "azureLogAnalytics": {
            "query": "customEvents \r\n| where $__timeFilter(timestamp)\r\n| where \"$var1\" == \"'All'\" or client_Browser == $var1\r\n| where \"$var2\" == \"'All'\" or client_City == $var2\r\n| summarize count() by client_Browser, client_City, bin(timestamp, 1d)\r\n| sort by timestamp asc",
            "resource": "/subscriptions/some-sub-guid/resourceGroups/some-rg/providers/microsoft.insights/components/some-app-insights-resource",
            "resultFormat": "time_series"
          },
          "azureMonitor": {
            "dimensionFilters": [],
            "timeGrain": "auto"
          },
          "datasource": {
            "type": "grafana-azure-monitor-datasource",
            "uid": "1Vz-WZ4nz"
          },
          "hide": false,
          "queryType": "Azure Log Analytics",
          "refId": "A",
          "subscription": "some-sub-guid"
        }
      ],
      "title": "Panel Title",
      "type": "timeseries"
    }
  ],
  "schemaVersion": 34,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": [
      {
        "allValue": "'All'",
        "current": {
          "selected": true,
          "text": "Edg 97.0",
          "value": "Edg 97.0"
        },
        "definition": "",
        "hide": 0,
        "includeAll": true,
        "multi": false,
        "name": "var1",
        "options": [],
        "query": {
          "azureLogAnalytics": {
            "query": "customEvents \r\n//| where $__timeFilter(timestamp)\r\n| distinct client_Browser",
            "resource": "/subscriptions/some-sub-guid/resourceGroups/some-rg/providers/microsoft.insights/components/some-app-insights-resource"
          },
          "queryType": "Azure Log Analytics",
          "refId": "A"
        },
        "refresh": 1,
        "regex": "",
        "skipUrlSync": false,
        "sort": 0,
        "type": "query"
      },
      {
        "allValue": "'All'",
        "current": {
          "selected": false,
          "text": "Redmond",
          "value": "Redmond"
        },
        "definition": "",
        "hide": 0,
        "includeAll": true,
        "multi": false,
        "name": "var2",
        "options": [],
        "query": {
          "azureLogAnalytics": {
            "query": "customEvents\r\n//| where $__timeFilter(timestamp)\r\n//| where client_Browser == \"$var1\"\r\n| distinct client_City",
            "resource": "/subscriptions/some-sub-guid/resourceGroups/some-rg/providers/microsoft.insights/components/some-app-insights-resource"
          },
          "queryType": "Azure Log Analytics",
          "refId": "A"
        },
        "refresh": 1,
        "regex": "",
        "skipUrlSync": false,
        "sort": 0,
        "type": "query"
      }
    ]
  },
  "time": {
    "from": "now-30d",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "test variables",
  "uid": "ab2fbabnz",
  "version": 17,
  "weekStart": ""
}
```

**Comment by mitulk** [comment]
@daniellee @hugohaggmark @shuotli

**Reply by mitulk** [reply]
@bradenhs on my team has submitted this PR (https://github.com/grafana/grafana/pull/46590) as a first step on this.

**Reply by mitulk** [reply]
@bradenhs has proposed a PR - https://github.com/grafana/grafana/pull/48218

**Reply by bradenhs** [reply]
I would love any feedback on the direction proposed in that PR! Would've happily put my comments in a GitHub issue/discussion but thought it'd be more useful to have some concrete working code alongside the proposal. Let me know if there is anything else I can do that'll make this proposal easier to discuss!

**Comment by joshhunt** [comment]
Hey @mitulk thanks for this write up. I've read through this a few times over the past few weeks and want to make sure I have a correct understanding.

The scenario is for having user-agent-like variables that contain Browser + Version. A single variable that lists all possible combination results to an explosion of values, you so split this into `$Browser` and `$Version` variables. When the user selects a `$Browser`, you want `$Version` to be updated with the valid versions for that browser, and when `$Version` is changed you want `$Browser` to be updated with valid versions for that `$Version`?

**Comment by mitulk** [comment]
Thanks for the response, @joshhunt.

You are correct! 

A few minor clarifications to help understand this a bit more:
1. Browser & Version is just an example. These are human readable, so I thought the example would be easy to understand. You could replace them with UUIDs which would be comparatively unreadable.
2. Version refers to Product Version, not version of Browser (they are different columns to begin with, so no splitting is needed)

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
